services/motion/builtin/replanner.go

```diff
@@ -11,40 +14,44 @@ type replanResponse struct {
 	replan bool
 }
 
+// replanFn is an alias for a function that will be polled by a replanner.
+type replanFn func(context.Context, [][]referenceframe.Input, int) (bool, error)
+
 // replanner bundles everything needed to execute a function at a given interval and return.
 type replanner struct {
 	period       time.Duration
-	fnToPoll     func(ctx context.Context) replanResponse
+	needReplan   replanFn
 	responseChan chan replanResponse
 }
 
 // newReplanner is a constructor.
-func newReplanner(period time.Duration, fnToPoll func(context.Context) replanResponse) *replanner {
+func newReplanner(period time.Duration, fnToPoll replanFn) *replanner {
 	return &replanner{
 		period:       period,
-		fnToPoll:     fnToPoll,
+		needReplan:   fnToPoll,
 		responseChan: make(chan replanResponse),
 	}
 }
 
 // startPolling executes the replanner's configured function at its configured period
 // The caller of this function should read from the replanner's responseChan to know when a replan is requested.
-func (r *replanner) startPolling(ctx context.Context) {
+func (r *replanner) startPolling(ctx context.Context, plan [][]referenceframe.Input, waypointIndex *atomic.Int32) {
 	ticker := time.NewTicker(r.period)
 	defer ticker.Stop()
-	for {
-		// this ensures that if the context is cancelled we always return early at the top of the loop
-		if ctx.Err() != nil {
-			return
-		}
 
+	// this check ensures that if the context is cancelled we always return early at the top of the loop
+	for ctx.Err() == nil {
 		select {
 		case <-ctx.Done():
 			return
 		case <-ticker.C:
-			response := r.fnToPoll(ctx)
-			if response.err != nil || response.replan {
-				r.responseChan <- response
+			index := int(waypointIndex.Load())
+			if index >= len(plan) {
```

Comment:
When would this happen? If this is happening, doesn't that mean that the execute function has a bug in incrementing the waypointIndex more than it should?

Reply:
Yeah thats true, I'll fix it in execute